apps/keyboard/js/imes/latin/worker.js

```diff
@@ -28,111 +31,282 @@
-      pendingPrediction.abort();
-
-    var noOfCandidates = prefix.length >= 2 && prefix.length <= 4 ?
-      CANDIDATES_HIGH :
-      CANDIDATES_LOW;
-
-    // Ask for 4 predictions, considering 50 candidates and considering
-    // only words with an edit distance of 1 (i.e. make only one correction
-    // per word)
-    pendingPrediction = Predictions.predict(prefix, 4, noOfCandidates, 1,
-                                            success, error);
-
-    function success(words) {
-      if (words.length > 0) {
-        // If the quality of the suggestions is very low, up candidates
-        if (prefix.length > 4 && words[0][1] < TOO_LOW_THRESHOLD) {
-          Predictions.predict(prefix, 4, CANDIDATES_HIGH, 1,
-                    function(words) {
-                      postMessage({ cmd: 'predictions',
-                                    input: prefix,
-                                    suggestions: words });
-                    }, error);
+    // Make sure we're not still running a previous one
+    if (pendingBuiltInPrediction) {
+      pendingBuiltInPrediction.abort();
+    }
+    if (pendingUserDictPrediction) {
+      pendingUserDictPrediction.abort();
+    }
+
+    // We merge predictions from built-in dictionaries and user dictionarys by:
+    // [1] For UserDict, ask for 4 predictions with 40 candidates, and
+    //     considering only words with an edit distance of 1 (i.e. make only one
+    //     correction per word)
+    // [2] For Built-In dict, everything remains the same (50 candidates if
+    //     prefix is short, 24 if not; do a second round:
+    //     [a] if there were results but results were bad; if prefix is long,
+    //         use 50 candidates.
+    //     [b] if there weren't results, use 60 candidates and allow 2
+    //         corrections.
+    //     As before weget 4 predictions.
+    // [3] Results from [1] and [2] are merged by:
+    //     [a] Results from [1] are not discarded even if quality is low.
+    //     [b] Results from [1] are annotated to come from user dict.
+    //     [c] Results are merged and re-sorted. Duplicated words (case- and
+    //         variants-sensitive) are eliminated by:
+    //         - Keeping the higher priority.
+    //         - Keeping the user dictionary flag.
+    //         The rationale is: latin.js will always show at least one user
+    //         dictionary word; since user specifically define some dict word,
+    //         he probably wants it very much.
+    // [4] latin.js will always retain at least one user dict prediction (unless
+    //     it decides to drop the prediction because it equals user input)
+
+    var BUILT_IN_CANDIDATES_LOW = 24;
+    var BUILT_IN_CANDIDATES_HIGH = 50;
+    var BUILT_IN_CANDIDATES_EXTRA = 60;
+    var USER_DICT_CANDIDATES = 40;
+    var BUILT_IN_TOO_LOW_THRESHOLD = 4;
+
+    var builtInPromise = new Promise(function(resolve, reject){
```

Comment:
Predictions.predict is sync, but the prediction process isn't. Wrapping both predictions in Promise means I can better synchronize things. We should discuss about this.